### ISSUE TITLE
Envoy does seemingly strange things with CORS. We should document this so people are not confused

### ISSUE THREAD

**Issue by nbkrause** [datawire/ambassador#2455]
If the plethora of blog posts about "how CORS works" is any indication, CORS is a confusing topic for many people. This can get even more confusing when you put Envoy in the mix.

Envoy has a `cors` `filter` that allows it to set the `Access-Control-Allow-*` headers based on configuration you give it (this is supported in Ambassador). 

A normal person would expect that is you set:
```
cors:
  origins:
  - http://foo.bar.com
```

That Envoy would set the `access-control-allow-origin: http://foo.bar.com` if a request comes in with `Origin: http://foo.bar.com` and not set that header otherwise.

Unfortunately, this is not exactly always the case. 

When Envoy set's these headers, it first checks to see if they have been set by the upstream service it proxied the request to. If they already exist, Envoy will not edit them and will return them, as is, to the client. If they are unset, Envoy will set them appropriately.

In practice, this behavior means that Envoy will default to the CORS configurations of the upstream service and provide a default if the upstream does not handle CORS. This is confusing and should be called out in the docs!

In summary: **If the upstream service set the `access-control-allow-*` headers, Envoy will not handle CORS as configured and will default to the upstream's CORS configuration**

This behavior is explained here https://github.com/envoyproxy/envoy/issues/300#issuecomment-268075987
